audit.md

```diff
@@ -0,0 +1,212 @@
+    - Low-severity: nice-to-haves, very low incentive attacks
+
+### Mid-severity
+
+- `AppProxy` should fail on construction if `_kernel` is not a contract
+    - `aragonOS/contracts/apps/AppProxy.sol`:
+        - Kernel-less apps are impossible: the fallback function will always fail on `kernel.getAppCode(appId)` if `kernel == 0` or if the kernel address is not a contract.
+    - `aragonOS/contracts/apps/AragonApp.sol`:
+        - `canPerform()` would not require the `address(kernel) == 0` check anymore as the proxy would fail to be created before that point
+    - `aragonOS/contracts/apm/ApmRegistry.sol`:
+        - `initialize()`'s `require(kernel != address(0));` check can be removed
+- `AppProxy` may use garbage data as its `target`:
+    - `aragonOS/contracts/apps/AppProxy.sol`:
+        - If the `AppProxy`'s `kernel` is set to a non-kernel contract address, `kernel.getAppCode(appId)` will return garbage data [due to how the fallback works](https://github.com/ethereum/solidity/issues/2630). This means that in `delegatedFwd(target, msg.data)`, `target` could potentially be delegating to a malicious contract. As far as I know, exploiting this would be extremely difficult, but it should be clear in the surrounding documentation that the `kernel` **MUST** be carefully set.
+- `KernelProxy` should fail on construction if `_kernelImpl` is not a contract
+    - `aragonOS/contracts/kernel/KernelProxy.sol`:
+        - `delegatedFwd(kernelImpl, msg.data)` will always fail if `kernelImpl == 0` or is not a contract. In such a case, it would also be impossible to run `upgradeKernel()`.
+- `KernelProxy`s' initialization can be front-runned:
+    - `aragonOS/contracts/kernel/KernelProxy.sol`:
+        - `initialize()` is public and accessible to anyone after the proxy is deployed. This should at least be highlighted in documentation for `Kernel.initialize()` and the `KernelProxy` so as to make sure users always create and initialize these proxies in one atomic transaction, but a better solution would be to introduce the same `initializePayload` that is in `AppProxy`.
+- `Kernel.revokePermission()` can be front-runned:
+    - `aragonOS/contracts/kernel/KernelProxy.sol`:
+        - There are many possible outcomes when `revokePermission()` races with an entity wishing to perform an action associated with the revoked role. As in all front-running cases, the entity learning of the revocation has the upper hand as he can send his transactions with much higher gas prices. I don't see any good way around this, as this is a pretty general problem plaguing all systems implementing revocable controls. It may be a good idea to note this in the comments and associated documentation. The best way around the problem is just to never give permissions to an entity you don't entirely trust, but this is obviously not very practical (yet).
```

Comment:
Or to introduce some artificial delays for entities you think could do that. I think it is a documentation + layer 2 solution, can't think of any elegant kernel level solutions.